patches/common/chromium/printing.patch

```diff
@@ -0,0 +1,1010 @@
++                                 bool is_modifiable,
++                                 const base::string16& device_name) {
+   DCHECK(task_runner_->RunsTasksInCurrentSequence());
+   DCHECK_EQ(page_number_, PageNumber::npos());
+
+@@ -200,6 +202,12 @@ void PrintJobWorker::GetSettings(bool ask_user_for_settings,
+             base::BindOnce(&PrintJobWorker::GetSettingsWithUI,
+                            base::Unretained(this), document_page_count,
+                            has_selection, is_scripted)));
++  } else if (!device_name.empty()) {
++    BrowserThread::PostTask(
++        BrowserThread::UI, FROM_HERE,
++        base::BindOnce(&WorkerHoldRefCallback, base::WrapRefCounted(query_),
++                       base::BindOnce(&PrintJobWorker::InitWithDeviceName,
++                                      base::Unretained(this), device_name)));
+   } else {
+     BrowserThread::PostTask(
+         BrowserThread::UI, FROM_HERE,
+@@ -316,6 +324,14 @@ void PrintJobWorker::UseDefaultSettings() {
+   GetSettingsDone(result);
+ }
+
++void PrintJobWorker::InitWithDeviceName(const base::string16& device_name) {
++  const auto& settings = printing_context_->settings();
++  std::unique_ptr<base::DictionaryValue> dic(new base::DictionaryValue);
++  printing::PrintSettingsToJobSettings(settings, dic.get());
++  dic->SetString(kSettingDeviceName, device_name);
++  UpdatePrintSettings(std::move(dic));
++}
++
+ void PrintJobWorker::StartPrinting(PrintedDocument* new_document) {
+   DCHECK(task_runner_->RunsTasksInCurrentSequence());
+
+diff --git a/chrome/browser/printing/print_job_worker.h b/chrome/browser/printing/print_job_worker.h
+index 182f7563e0c7..eed2230627d6 100644
+--- a/chrome/browser/printing/print_job_worker.h
++++ b/chrome/browser/printing/print_job_worker.h
+@@ -54,7 +54,8 @@ class PrintJobWorker {
+                    bool has_selection,
+                    MarginType margin_type,
+                    bool is_scripted,
+-                   bool is_modifiable);
++                   bool is_modifiable,
++                   const base::string16& device_name);
+
+   // Set the new print settings from a dictionary value.
+   void SetSettings(std::unique_ptr<base::DictionaryValue> new_settings);
+@@ -155,6 +156,9 @@ class PrintJobWorker {
+   // systems.
+   void UseDefaultSettings();
+
++  // set the printer name
++  void InitWithDeviceName(const base::string16& device_name);
++
+   // Printing context delegate.
+   const std::unique_ptr<PrintingContext::Delegate> printing_context_delegate_;
+
+diff --git a/chrome/browser/printing/print_preview_message_handler.cc b/chrome/browser/printing/print_preview_message_handler.cc
+index e1613a9b52ad..e10959772a39 100644
+--- a/chrome/browser/printing/print_preview_message_handler.cc
++++ b/chrome/browser/printing/print_preview_message_handler.cc
+@@ -35,6 +35,8 @@
+ #include "printing/print_job_constants.h"
+ #include "printing/print_settings.h"
+
++#include "atom/common/node_includes.h"
```

Comment:
this is a little terrifying, what's it needed for?

Reply:
If I remember correctly, it's needed [here](https://github.com/electron/electron/pull/15023/files/cb0cc4d9359d6fe2036874dd3f5a3667663cb8a3#diff-eec8ce3c51ad49f522e2bf2bdc6a9bafR241)

Reply:
ahh. okay then. ideally we'd keep all interaction with node.js under `electron/` (this would be the first patch that adds a node dependency to a chromium file). would it be possible to refactor that `RunPrintToPDFCallback` function to kick the ball back to something in Electron, which could then do the node buffer creation?

... alternatively, is it possible that node buffers and v8 ArrayBuffers are actually one and the same thing? in that case, would you be able to just create a v8 buffer here instead of pulling in node as a dependency?